dora/src/cannon/codegen.rs

```diff
@@ -1575,38 +1574,35 @@ where
         reg
     }
 
-    fn emit_invoke_direct_void(&mut self, fct_def_id: FctDefId, start_reg: Register, num: u32) {
-        self.emit_invoke_direct(fct_def_id, start_reg, num, None);
+    fn emit_invoke_direct_void(&mut self, fct_def_id: FctDefId, num: u32) {
+        self.emit_invoke_direct(fct_def_id, num, None);
     }
 
-    fn emit_invoke_direct_generic(
-        &mut self,
-        dest: Register,
-        fct_id: FctDefId,
-        start_reg: Register,
-        num: u32,
-    ) {
+    fn emit_invoke_direct_generic(&mut self, dest: Register, fct_id: FctDefId, num: u32) {
         let bytecode_type = self.bytecode.register_type(dest);
 
-        let reg = self.emit_invoke_direct(fct_id, start_reg, num, Some(bytecode_type));
+        let reg = self.emit_invoke_direct(fct_id, num, Some(bytecode_type));
 
         self.emit_store_register(reg, dest);
     }
 
     fn emit_invoke_direct(
         &mut self,
         fct_def_id: FctDefId,
-        start_reg: Register,
         num: u32,
         bytecode_type: Option<BytecodeType>,
     ) -> AnyReg {
         assert!(num > 0);
 
-        let bytecode_type_self = self.bytecode.register_type(start_reg);
+        assert_eq!(self.argument_stack.len() as u32, num);
+        let arguments = self.argument_stack.drain(..).collect::<Vec<_>>();
```

Comment:
I think you can take ownership of the vector (and replace the current one with an empty one) with: `let arguments = mem::replace(&mut self.argument_stack, Vec::new())`.